File: Src/CameraServerApp/CameraServer/VideoReader/cv2Reader.py
from CameraServer.common.logger import _LOGGER

import cv2
import os

class camera_server():
    def __init__(self, video) -> None:
        _LOGGER.info(video)
        self.cap = cv2.VideoCapture(video)
        if self.cap.isOpened():
            # Video is successfully opened
            _LOGGER.info("Video successfully read")
        else:
            # Failed to open the video
            _LOGGER.error("Failed to read video")
            exit()

    def readVideo(self):
        try:
            ret, frame = self.cap.read()
        except:
            _LOGGER.error("Something wrong here")
            exit(0)

        if ret == False:
            self.cap.set(cv2.CAP_PROP_POS_FRAMES, 0)
            ret, frame = self.cap.read()
        frame = cv2.resize(frame, (0,0), fx=0.5, fy=0.5)
        return frame
    
    def getCaptureInformation(self, parameters : list) -> dict:
        metadata = {}
        for parameter in parameters:
            match parameter:
                case 'height':
                    info = self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT)
                case 'width':
                    info = self.cap.get(cv2.CAP_PROP_FRAME_WIDTH)
                case 'fps':
                    info = self.cap.get(cv2.CAP_PROP_FPS)
            metadata[parameter] = info
        return metadata
    
    def isOpen(self) -> bool:
        return self.cap.isOpened()
    
    def close(self):
        self.cap.release()

File: Src/CameraServerApp/cameraserver/src/main/java/com/huy/VideoReader/cv2Reader.java
package com.huy.VideoReader;

import com.huy.App;
import com.huy.Shared.helper;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.opencv.core.Size;
import org.opencv.videoio.VideoCapture;
import org.opencv.videoio.Videoio;
import org.opencv.imgproc.Imgproc;
import org.opencv.core.Mat;
import org.opencv.imgcodecs.Imgcodecs;

import java.util.HashMap;
import java.util.Map;


public class cv2Reader {
    private static final Logger LOGGER = LogManager.getLogger(cv2Reader.class);
    private final VideoCapture cap;
    public cv2Reader(String filename){
        cap = new VideoCapture(filename, Videoio.CAP_FFMPEG);
        if (!cap.isOpened()){
            LOGGER.error("Open video failed");
            System.exit(0);
        }
        LOGGER.info("Open video success");
    }

    public final Mat readVideo(){
        Mat frame = new Mat();
        if (!cap.read(frame)){
            cap.set(Videoio.CAP_PROP_POS_FRAMES, 0);
            cap.read(frame);
        }
        return frame;
    }
    public Map<String, String> getCaptureInformation(String... args){
        Map<String, String> metadata = new HashMap<>();
        String info = "";
        for (String arg : args){
            switch (arg){
                case "height":
                    info = Double.toString(cap.get(Videoio.CAP_PROP_FRAME_HEIGHT));
                    break;
                case "width":
                    info = Double.toString(cap.get(Videoio.CAP_PROP_FRAME_WIDTH));
                    break;
                case "fps":
                    info = Double.toString(cap.get(Videoio.CAP_PROP_FPS));
                    break;
                default:
                    break;
            }
            metadata.put(arg, info);
        }
        return  metadata;
    }

    public final void close(){
        cap.release();
    }
}
